File: LuaJEngine/src/main/java/org/anime_game_servers/luaj_engine/LuaJValue.java
package org.anime_game_servers.luaj_engine;

import org.anime_game_servers.lua.engine.LuaValue;

public class LuaJValue implements LuaValue {
    private final org.luaj.vm2.LuaValue value;
    private final LuaJEngine engine;

    LuaJValue(LuaJEngine engine, org.luaj.vm2.LuaValue value) {
        this.engine = engine;
        this.value = value;
    }

    @Override
    public boolean isNull() {
        return value.isnil();
    }

    @Override
    public boolean isBoolean() {
        return value.isboolean();
    }

    @Override
    public boolean isInteger() {
        return value.isint();
    }

    @Override
    public boolean isLong() {
        return value.islong();
    }

    @Override
    public boolean isDouble() {
        return value.isnumber();
    }

    @Override
    public boolean isFloat() {
        return value.isnumber();
    }

    @Override
    public boolean isString() {
        return value.isstring();
    }

    @Override
    public boolean isTable() {
        return value.istable() || value.isuserdata();
    }

    @Override
    public boolean asBoolean() {
        return value.toboolean();
    }

    @Override
    public int asInteger() {
        return value.toint();
    }

    @Override
    public long asLong() {
        return value.tolong();
    }

    @Override
    public double asDouble() {
        return value.todouble();
    }

    @Override
    public float asFloat() {
        return value.tofloat();
    }

    @Override
    public String asString() {
        return value.toString();
    }

    @Override
    public <T> T asObject(Class<T> type) {
        if (!value.istable() && !value.isuserdata()) {
            return null;
        }

        if(value.isuserdata()) {
            return (T) value.checkuserdata(type);
        }

        return engine.getSerializer().toObject(type, value.checktable());
    }
}

File: LuaJEngine/src/main/java/org/anime_game_servers/luaj_engine/LuaJValue.kt
package org.anime_game_servers.luaj_engine

import org.luaj.vm2.LuaValue

class LuaJValue internal constructor(private val engine: LuaJEngine, private val value: LuaValue) :
    org.anime_game_servers.lua.engine.LuaValue {
    override fun isNull(): Boolean {
        return value.isnil()
    }

    override fun isBoolean(): Boolean {
        return value.isboolean()
    }

    override fun isInteger(): Boolean {
        return value.isint()
    }

    override fun isLong(): Boolean {
        return value.islong()
    }

    override fun isDouble(): Boolean {
        return value.isnumber()
    }

    override fun isFloat(): Boolean {
        return value.isnumber()
    }

    override fun isString(): Boolean {
        return value.isstring()
    }

    override fun isTable(): Boolean {
        return value.istable() || value.isuserdata()
    }

    override fun asBoolean(): Boolean {
        return value.toboolean()
    }

    override fun asInteger(): Int {
        return value.toint()
    }

    override fun asLong(): Long {
        return value.tolong()
    }

    override fun asDouble(): Double {
        return value.todouble()
    }

    override fun asFloat(): Float {
        return value.tofloat()
    }

    override fun asString(): String {
        return value.toString()
    }

    override fun <T> asObject(type: Class<T>): T? {
        if (!value.istable() && !value.isuserdata()) {
            return null
        }

        if (value.isuserdata()) {
            return value.checkuserdata(type) as T?
        }

        return engine.serializer.toObject<T?>(type, value.checktable())
    }

    override fun <K, V> asMap(
        keyType: Class<K>,
        valueType: Class<V>
    ): Map<K, V>? {
        if (!value.istable()) {
            return null
        }

        return engine.serializer.toMap(keyType, valueType, value.checktable())
    }

    override fun <T> asList(type: Class<T>): List<T>? {
        if (!value.istable()) {
            return null
        }

        return engine.serializer.toList(type, value.checktable())
    }
}
